design/local-node-agent.md

```diff
@@ -0,0 +1,123 @@
+
+#### Improvements Over Rook's Current Support for Persistent Volumes
+Rook currently has a hybrid approach of providing persistent storage in a Kubernetes cluster.  While the Rook operator implements a dynamic volume provisioner, the attach and mounting is offloaded to the existing [RBD plugin](https://github.com/kubernetes/kubernetes/tree/master/pkg/volume/rbd).
+This requires that the Ceph tools are installed alongside the Kubelet on every node in the cluster, which causes friction in the experience for users of Rook.
+Furthermore, if users want to consume a Rook volume outside of the default namespace, then they must manually copy secrets to their namespace of choice.
+
+Both of these issues are handled by this proposed design for the Rook agent, greatly streamlining the Rook storage experience in Kubernetes and removing the common causes for errors and failure.  This is a big win for our users.
+
+Also, this proposed design allows Rook to normalize on a single path for providing persistent volumes across all Kubernetes versions that it supports (1.5+).
+
+#### Fencing
+Ensuring exclusive single client access (`ReadWriteOnce`) to the persistent volume can be achieved by the Rook agent, independently from the underlying storage protocol that is being used (RBD, NBD, iSCSI, etc.).
+Remember that the attach operation will be initiated on the desired node's Rook agent by the creation of a volume attach CRD.
+When the agent performs the attach on its local node, it can update the CRD with information stating that the volume is now locked.
+Thus, the CRD itself is providing the accounting for the volume's lock in a generalized way, independent of the of the underlying storage protocol the agent chose to use for the volume.
+Of course, this lock can be augmented at the lower layer of the specific storage protocol, e.g., `rbd lock add` for RBD.
+
+Similarly, during the detach operation, the CRD will be deleted to signify that the volume has been unlocked.
+Since the Rook Flexvolume driver implements the attach/detach interface, detach will be called on the Controller manager node.
+If the node where the volume has attached has died, then this central driver/agent will need to perform the unlocking itself.
+This can be done by executing the storage protocol specific command (e.g., `rbd lock remove`) and the deletion of the volume attach CRD.
+
+#### Security
+There are two interfaces for communicating with and invoking operations on the Rook agent:
+1. Unix domain socket
+1. CRD events
+
+The Unix domain socket will have read/write only accessible by `root` and it is only accessible on the local node (not remotely accessible).
+
+CRD events (such as the creation of the volume attach CRD) will be using the standard Kubernetes security enforcement with RBAC.  Only cluster roles that are explicitly granted permissions to `GET`, `CREATE`, `DELETE`, etc. will be allowed to do so.
+
+### Cluster Cleanup
```

Comment:
another cluster cleanup task could be to clean up `dataDirHostPath`.  #927 made me think of this.